apps/cowswap-frontend/src/modules/appData/utils/buildAppData.ts

```diff
@@ -125,12 +138,61 @@ function mergeHooks(
   if (!hooks1 && hooks2) return hooks2
 
   if (hooks1 && hooks2) {
+    const pre = (hooks1.pre || []).concat(hooks2.pre || [])
+    const post = (hooks1.post || []).concat(hooks2.post || [])
+
+    if (!pre && !post) {
+      // Avoid empty hooks
+      console.log(`bug:mergeHooks empty pre and post`, hooks1, hooks2)
+      return undefined
+    }
+
+    console.log(`bug:mergeHooks has either pre, post or both`, hooks1, hooks2, pre, post)
+
     return {
       version: hooks1.version,
-      pre: (hooks1.pre || []).concat(hooks2.pre || []),
-      post: (hooks1.post || []).concat(hooks2.post || []),
+      // Remove the ones that are empty here too
+      ...(pre ? { pre } : undefined),
+      ...(post ? { post } : undefined),
     }
   }
 
   return undefined
 }
+
+function removeDuplicatedHook<T extends PreHooks | PostHooks>(hooks: T | undefined): T | undefined {
+  if (!hooks || hooks.length < 2) {
+    console.log(`bug:removeDuplicatedHook no hooks or only 1`, hooks?.length)
+    return hooks
+  }
+
+  const duplicatedIndices: Set<number> = new Set()
+
+  // Check all hooks against each other to identify duplicates
+  for (let i = 0; i < hooks.length; i++) {
+    if (duplicatedIndices.has(i)) {
+      console.log(`bug:removeDuplicatedHook skipping duplicate`, i)
+      continue
+    }
+
+    const hookA = hooks[i]
+
+    for (let j = i + 1; j < hooks.length; j++) {
+      const hookB = hooks[j]
+
+      if (hooksAreEqual(hookA, hookB)) {
+        console.log(`bug:removeDuplicatedHook found duplicate`, hookB)
+        duplicatedIndices.add(j)
+      }
+    }
+  }
+
+  return duplicatedIndices.size ? (hooks.filter((hook, index) => !duplicatedIndices.has(index)) as T) : hooks
+}
+
+function hooksAreEqual(hookA: CowHook, hookB: CowHook): boolean {
+  return (
+    hookA.callData.toLowerCase() === hookB.callData.toLowerCase() &&
+    hookA.target.toLowerCase() === hookB.target.toLowerCase()
+  )
+}
```

Comment:
The case I found so far is when both callData and target were the same.
Can similar hooks with different values happen?

Reply:
No, I don't think makes functional sense, so probaly is right to now allow. Just saying that for the hooks we create we might want to be more restrictive (never allow two permit hooks in the app-date created by our app). 
If the amount of the permit is different (i.e. somehow the user manage to sign two permits with different amount for the same sell token) then we would discard one (the one with the smaller allowance ideally)